libmcount/mcount.c

```diff
@@ -1237,12 +1237,20 @@ static int __mcount_entry(unsigned long *parent_loc, unsigned long child,
 	rstack->nr_events  = 0;
 	rstack->event_idx  = ARGBUF_SIZE;
 
-	/* hijack the return address of child */
-	*parent_loc = mcount_return_fn;
+	if (unlikely(mcount_flat)) {
+		/* record function entry only in flat mode */
+		mtdp->idx = 0;
+		rstack->depth = 0;
+		record_trace_data(mtdp, rstack, NULL);
```

Comment:
This should be called in `mcount_entry_filter_record()` instead.  How about this?

```diff
diff --git a/libmcount/mcount.c b/libmcount/mcount.c
index a471fad3..96441ce2 100644
--- a/libmcount/mcount.c
+++ b/libmcount/mcount.c
@@ -1021,7 +1021,7 @@ void mcount_entry_filter_record(struct mcount_thread_data *mtdp,
                        if (mcount_watchpoints)
                                save_watchpoint(mtdp, rstack, mcount_watchpoints);
 
-                       if (mtdp->nr_events) {
+                       if (mtdp->nr_events || mcount_flat) {
                                bool flush = false;
                                int i;
 
@@ -1033,7 +1033,7 @@ void mcount_entry_filter_record(struct mcount_thread_data *mtdp,
                                        if (mtdp->event[i].idx == ASYNC_IDX)
                                                flush = true;
 
-                               if (flush)
+                               if (flush || mcount_flat)
                                        record_trace_data(mtdp, rstack, NULL);
                        }
                }
```